easybuild/easyblocks/f/flang.py

```diff
@@ -0,0 +1,186 @@
+        for tmp in self.src:
+            if tmp['name'].startswith("flang-llvm-"):
+                self.llvm_src_dir = tmp['finalpath']
+                break
+
+        if self.llvm_src_dir is None:
+            raise EasyBuildError(
+                "Could not determine LLVM source root (LLVM source was not unpacked?)")
+        src_dirs = {}
+
+        def find_source_dir(globpatterns, targetdir):
+            """Search for directory with globpattern and rename it to targetdir"""
+            if not isinstance(globpatterns, list):
+                globpatterns = [globpatterns]
+
+            glob_src_dirs = [glob_dir for globpattern in globpatterns for glob_dir in
+                             glob.glob(globpattern)]
+            if len(glob_src_dirs) != 1:
+                raise EasyBuildError(
+                    "Failed to find exactly one source directory for pattern %s: %s",
+                    globpatterns,
+                    glob_src_dirs)
+            src_dirs[glob_src_dirs[0]] = targetdir
+
+        if LooseVersion(self.version) >= LooseVersion('3.8'):
+            find_source_dir('openmp-*',
+                            os.path.join(self.llvm_src_dir, 'projects', 'openmp'))
+
+        find_source_dir('flang-driver-*',
+                        os.path.join(self.llvm_src_dir, 'tools', 'clang'))
+
+        # Place the flang code in a separate directory for the build step for after
+        # we've built llvm
+        self.flang_source_dir = os.path.join(self.llvm_src_dir, '..', 'flang')
+        find_source_dir('flang-flang_*', self.flang_source_dir)
+
+        for src in self.src:
+            for (dirname, new_path) in src_dirs.items():
+                if src['name'].startswith(dirname):
+                    old_path = os.path.join(src['finalpath'], dirname)
+                    try:
+                        shutil.move(old_path, new_path)
+                    except IOError as err:
+                        raise EasyBuildError("Failed to move %s to %s: %s", old_path,
+                                             new_path, err)
+                    src['finalpath'] = new_path
+                    break
+
+    def build_with_temporary_llvm(self, build_dir, src_dir, parallel=True, additional_options=[]):
+        """Build Clang stage N using Clang stage N-1"""
+
+        # Create and enter build directory.
+        mkdir(build_dir)
+        change_dir(build_dir)
+
+        # Configure.
+        CC = os.path.join(self.llvm_obj_dir, 'bin', 'clang')
+        CXX = os.path.join(self.llvm_obj_dir, 'bin', 'clang++')
+        FC = os.path.join(self.llvm_obj_dir, 'bin', 'flang')
+        LLVM_CONFIG = os.path.join(self.llvm_obj_dir, 'bin', 'llvm-config')
+        shlib_ext = get_shared_lib_ext()
+        LIBOMP = os.path.join(self.llvm_obj_dir, 'lib', 'libomp.%s' % shlib_ext)
+
+        options = "-DCMAKE_INSTALL_PREFIX=%s " % self.installdir
+        options += "-DCMAKE_C_COMPILER='%s' " % CC
```

Comment:
This could be cleaned up quite a bit...

```python
opts_map = {
    'CMAKE_INSTALL_PREFIX': self.installdir,
    'CMAKE_C_COMPILER': cc,
}
options = ' '.join('-D%s=%s' % item for item in sorted(opts_map.items()))
options += ' ' + ' '.join(additional_options) + ' ' + self.cfg['configopts']